cmd/mdatagen/templates/match.go.tmpl

```diff
@@ -0,0 +1,234 @@
+// Code generated by mdatagen. DO NOT EDIT.
+
+package metadata
+
+import (
+	"errors"
+	"io"
+	"sort"
+	"strings"
+)
+
+var ErrNotMultimatch = errors.New("this index doesn't represent a valid multimatch pattern")
+
+type Pattern struct {
+	pattern     string
+	level       int
+	hasWildcard bool
+	valueAssign any
+}
+
+func NewPattern(pattern string, valueAssign any) Pattern {
+	level := 0
+	for _, c := range pattern {
+		if c == '.' {
+			level++
+		}
+	}
+	wildcard := strings.Contains(pattern, "*")
+	return Pattern{
+		pattern:     pattern,
+		level:       level,
+		hasWildcard: wildcard,
+		valueAssign: valueAssign,
+	}
+}
+
+func (p Pattern) Match(s string) bool {
+	if len(s) == 0 {
+		return true
+	}
+	lastWildcard := -1
+	patternScan := &scanner{str: p.pattern}
+	strScan := &scanner{str: s}
+	for !strScan.isFinished() {
+		strChar, err := strScan.current()
+		if err != nil {
+			return false
+		}
+		patternChar, err := patternScan.current()
+		if err != nil {
+			return false
+		}
+
+		switch patternChar {
+		case '*':
+			if patternScan.isLast() {
+				return true
+			}
+			patternNext, err := patternScan.peek()
+			if err == nil && strChar == patternNext {
+				lastWildcard = patternScan.idx
+				patternScan.idx += 2
+			}
+
+		case '{':
```

Comment:
Can you please point to some specifications for this sort of pattern matching? Does it usually go with wildcard matching?

Reply:
Done in [d82df38](https://github.com/open-telemetry/opentelemetry-collector/pull/10065/commits/d82df38d422314a454be87805bd2f69bde76b472)